File: LeetCode - Algorithm/Binary Tree/230.java
// 230. Kth Smallest Element in a BST
// https://leetcode.com/problems/kth-smallest-element-in-a-bst/

import java.util.List;
import java.util.ArrayList;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    private List<Integer> values;

    Solution() {
        this.values = new ArrayList<>();
    }

    public int kthSmallest(TreeNode root, int k) {
        this.dfsInOrder(root);

        return this.values.get(k - 1);
    }

    public void dfsInOrder(TreeNode node) {
        if (node == null)
            return;

        this.dfsInOrder(node.left);
        this.values.add(node.val);
        this.dfsInOrder(node.right);
    }
}

File: LeetCode - Algorithm/Binary Tree/230.js
// 230. Kth Smallest Element in a BST
// https://leetcode.com/problems/kth-smallest-element-in-a-bst/

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function (root, k) {
  const values = [];

  getValues(root, values, k);

  return values[k - 1];
};

const getValues = (node, values, target) => {
  if (!node) return;

  if (values.length >= target) return;

  getValues(node.left, values, target);

  if (values.length >= target) return;

  values.push(node.val);

  if (values.length >= target) return;

  getValues(node.right, values, target);
};
